File: services/api/src/cvops_api/schemas/samples.py
from __future__ import annotations
import uuid
from datetime import datetime
from typing import Any, Generic, Literal, TypeVar
from pydantic import BaseModel, Field, model_validator

T = TypeVar("T")


class CursorPage(BaseModel, Generic[T]):
    items: list[T]
    next_cursor: str | None = None


class TagBrief(BaseModel):
    """Minimal tag shape embedded in a SampleOut."""

    model_config = {"from_attributes": True}
    id: uuid.UUID
    name: str
    color: str


class SampleOut(BaseModel):
    model_config = {"from_attributes": True, "populate_by_name": True}
    id: uuid.UUID
    project_id: uuid.UUID
    blob_hash: str
    source_id: uuid.UUID
    width: int
    height: int
    frame_index: int | None = None
    perceptual_hash: str | None = None
    # Reads the ORM's `metadata_` attribute but serializes as the clean `metadata`
    # (matches the frontend Sample type and avoids leaking the trailing underscore).
    metadata: dict[str, Any] | None = Field(
        None, validation_alias="metadata_", serialization_alias="metadata"
    )
    review_status: str = "unreviewed"
    created_at: datetime
    # Enriched per-page by core.sample_view.build_sample_outs (not plain ORM columns).
    tags: list[TagBrief] = []
    has_annotations: bool = False
    latest_revision_id: uuid.UUID | None = None


class SampleUpdate(BaseModel):
    """Edit a sample's mutable operational fields. Content (blob/dims) is immutable."""

    metadata: dict[str, Any] | None = None
    metadata_mode: Literal["merge", "replace"] = "merge"
    tag_ids: list[uuid.UUID] | None = None


BulkAction = Literal["delete", "set_review_status", "add_tags", "remove_tags", "add_to_collection"]

REVIEW_STATUSES = ("unreviewed", "accepted", "rejected")


class SampleBulkAction(BaseModel):
    """Unified bulk action over a set of samples (all scoped to one project)."""

    action: BulkAction
    sample_ids: list[uuid.UUID]
    review_status: str | None = None
    tag_ids: list[uuid.UUID] | None = None
    collection_id: uuid.UUID | None = None

    @model_validator(mode="after")
    def _check_required(self) -> "SampleBulkAction":
        if self.action == "set_review_status":
            if self.review_status not in REVIEW_STATUSES:
                raise ValueError(f"review_status must be one of {REVIEW_STATUSES}")
        elif self.action in ("add_tags", "remove_tags"):
            if not self.tag_ids:
                raise ValueError(f"tag_ids is required for action={self.action}")
        elif self.action == "add_to_collection":
            if self.collection_id is None:
                raise ValueError("collection_id is required for action=add_to_collection")
        return self


class BulkResult(BaseModel):
    matched: int
    affected: int
    skipped_ids: list[uuid.UUID] = []


class SampleIdList(BaseModel):
    sample_ids: list[uuid.UUID]


class TagIdList(BaseModel):
    tag_ids: list[uuid.UUID]


class AnnotationRevisionOut(BaseModel):
    model_config = {"from_attributes": True}
    id: uuid.UUID
    sample_id: uuid.UUID
    ontology_id: uuid.UUID
    ontology_version: int
    revision_no: int
    # payload is the list of annotation objects ([{class_key, geometry}, ...]),
    # not a dict — typing it as a dict made response validation 500.
    payload: list[dict[str, Any]]
    provenance: dict[str, Any] | None = None
    created_at: datetime


class AnnotationCreate(BaseModel):
    ontology_id: uuid.UUID
    payload: dict[str, Any]
    provenance: dict[str, Any] | None = None

File: services/frontend/src/api/samples.ts
import { useInfiniteQuery, useMutation, useQuery, useQueryClient } from '@tanstack/react-query'
import { client } from '../lib/client'
import { PRESIGNED_URL_GC_MS, PRESIGNED_URL_STALE_MS } from '../lib/presign'

export interface TagBrief {
  id: string
  name: string
  color: string
}

export interface Sample {
  id: string
  project_id: string
  blob_hash: string
  source_id: string
  width: number
  height: number
  frame_index: number | null
  perceptual_hash: string | null
  metadata: Record<string, unknown> | null
  review_status: string
  tags: TagBrief[]
  has_annotations: boolean
  latest_revision_id: string | null
  created_at: string
}

export interface CursorPage<T> {
  items: T[]
  next_cursor: string | null
}

export interface SampleFilters {
  source_id?: string
  review_status?: string
  has_annotations?: boolean
  annotation_class?: string
  collection_id?: string
  tag_id?: string
  created_after?: string
  created_before?: string
}

function filterParams(filters: SampleFilters): URLSearchParams {
  const params = new URLSearchParams({ limit: '50' })
  for (const [k, v] of Object.entries(filters)) {
    if (v !== undefined && v !== '') params.set(k, String(v))
  }
  return params
}

export function useSamples(projectId: string | undefined, filters: SampleFilters = {}) {
  return useInfiniteQuery<CursorPage<Sample>>({
    queryKey: ['samples', projectId, filters],
    queryFn: async ({ pageParam }) => {
      const params = filterParams(filters)
      if (pageParam) params.set('cursor', pageParam as string)
      const { data } = await client.get<CursorPage<Sample>>(
        `/projects/${projectId}/samples?${params}`,
      )
      return data
    },
    initialPageParam: null,
    getNextPageParam: (last) => last.next_cursor ?? undefined,
    enabled: !!projectId,
  })
}

export function useThumbnailUrl(sampleId: string | undefined) {
  return useQuery<{ url: string }>({
    queryKey: ['thumbnail', sampleId],
    queryFn: async () => {
      const { data } = await client.get<{ url: string }>(`/samples/${sampleId}/thumbnail-url`)
      return data
    },
    enabled: !!sampleId,
    staleTime: PRESIGNED_URL_STALE_MS,
    gcTime: PRESIGNED_URL_GC_MS,
  })
}

export function useImageUrl(sampleId: string | undefined) {
  return useQuery<{ url: string }>({
    queryKey: ['image-url', sampleId],
    queryFn: async () => {
      const { data } = await client.get<{ url: string }>(`/samples/${sampleId}/image-url`)
      return data
    },
    enabled: !!sampleId,
    staleTime: PRESIGNED_URL_STALE_MS,
    gcTime: PRESIGNED_URL_GC_MS,
  })
}

export interface SampleUpdate {
  metadata?: Record<string, unknown>
  metadata_mode?: 'merge' | 'replace'
  tag_ids?: string[]
}

export function usePatchSample(projectId: string | undefined) {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async (vars: { id: string; body: SampleUpdate }) => {
      const { data } = await client.patch<Sample>(`/samples/${vars.id}`, vars.body)
      return data
    },
    onSuccess: (data) => {
      qc.invalidateQueries({ queryKey: ['samples', projectId] })
      qc.invalidateQueries({ queryKey: ['sample', data.id] })
    },
  })
}

export function useDeleteSample(projectId: string | undefined) {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async (id: string) => {
      await client.delete(`/samples/${id}`)
    },
    onSuccess: () => qc.invalidateQueries({ queryKey: ['samples', projectId] }),
  })
}

export type BulkAction =
  | 'delete'
  | 'set_review_status'
  | 'add_tags'
  | 'remove_tags'
  | 'add_to_collection'

export interface SampleBulkAction {
  action: BulkAction
  sample_ids: string[]
  review_status?: string
  tag_ids?: string[]
  collection_id?: string
}

export interface BulkResult {
  matched: number
  affected: number
  skipped_ids: string[]
}

export function useBulkSampleAction(projectId: string | undefined) {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async (body: SampleBulkAction) => {
      const { data } = await client.post<BulkResult>(
        `/projects/${projectId}/samples/bulk`,
        body,
      )
      return data
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['samples', projectId] })
      qc.invalidateQueries({ queryKey: ['collections', projectId] })
    },
  })
}
